### PR TITLE
Improve SQL PGDS data source configs

### PR DESCRIPTION
**Pull request by tobias-johansson**
Rework the data source configs
- Specific case classes for different data source types enabling
  - Better config scala API
  - Early validation of configuration
- Update examples to show API config over JSON config
- Make json representation a map instead of a list

Scala api:
```scala
GraphSources
  .sql("my.ddl")
  .withSqlDataSourceConfigs(
    "CENSUS_ORACLE" -> Jdbc(
      url = "<jdbc url>",
      driver = "org.h2.Driver"
    ),
    "ORACLE_X2" -> Jdbc(
      url = "<jdbc url>",
      driver = "org.h2.Driver",
      options = Map(
        "user" -> "ORACLE_X2_USER",
        "password" -> "ORACLE_X2_PASS",
        "fetchSize" -> "10"
      )
    ),
    "HIVE_CENSUS" -> Hive(),
    "HIVE_X2" -> Hive(),
    "MY_DIR" -> File(
      format = ParquetFormat,
      basePath = Some("/my/path")
    )
  )
```

Json now looks like:
```json
{
  "CENSUS_ORACLE": {
    "type": "jdbc",
    "url": "<jdbc url>",
    "driver": "org.h2.Driver"
  },

  "ORACLE_X2": {
    "type": "jdbc",
    "url": "<jdbc url>",
    "driver": "org.h2.Driver",
    "options": {
      "user": "ORACLE_X2_USER",
      "password": "ORACLE_X2_PASS",
      "fetchSize": "10"
    }
  },

  "HIVE_CENSUS": {
    "type": "hive"
  },

  "HIVE_X2": {
    "type": "hive"
  },

  "MY_DIR": {
    "type": "file",
    "format": "parquet",
    "basePath": "/my/path"
  }
}
```

### CODE REVIEW

File: spark-cypher-testing/src/main/scala/org/opencypher/spark/testing/utils/H2Utils.scala [hunk 1]
```diff
@@ -45,18 +45,30 @@ object H2Utils {
...
+      case (Some(user), Some(pass)) =>
+        DriverManager.getConnection(cfg.jdbcUri.get, user, pass)
+      case _ =>
+        DriverManager.getConnection(cfg.jdbcUri.get)
+    }
     try { code(conn) } finally { conn.close() }
   }
 
-  implicit class DataframeSqlOps(df: DataFrame) {
+  implicit class DataFrameReaderOps(write: DataFrameWriter[Row]) {
```
Comment (s1ck) [inline spark-cypher-testing/src/main/scala/org/opencypher/spark/testing/utils/H2Utils.scala]: `DataFrame*Writer*Ops` sounds better :>

File: spark-cypher-testing/src/test/resources/sql/sql-data-sources.json [hunk 2]
```diff
@@ -13,6 +13,8 @@
     "schemaName": "X2",
     "jdbcUri": "jdbc:h2:mem:X2.db;INIT=CREATE SCHEMA IF NOT EXISTS X2;DB_CLOSE_DELAY=30;",
     "jdbcDriver": "org.h2.Driver",
+    "jdbcUser": "ORACLE_X2_USER",
```
Comment (s1ck) [inline spark-cypher-testing/src/test/resources/sql/sql-data-sources.json]: Can we still put user/pw into the URI? If user/pw is defined in both places, is there an exception or do we prioritize?

Comment (s1ck) [reply spark-cypher-testing/src/test/resources/sql/sql-data-sources.json]: Maybe add one entry that mirrors that behavior and add a test?

**Comment by s1ck** [comment]
Seems I need to re-review

File: spark-cypher/src/main/scala/org/opencypher/spark/api/io/sql/SqlDataSourceConfig.scala [hunk 3]
```diff
@@ -26,50 +26,86 @@
...
+  private implicit val hive: ReadWriter[Hive.type] = macroRW
+  private implicit val file: ReadWriter[File] = macroRW
+  private val defaultMacroRW: ReadWriter[SqlDataSourceConfig] = macroRW
 
-  implicit def rw: ReadWriter[SqlDataSourceConfig] = macroRW
+  private final val UJSON_TYPE_KEY = "$type"
+  private final val CAPS_TYPE_KEY = "type"
 
-  def toJson(dataSource: SqlDataSourceConfig): String = write[SqlDataSourceConfig](dataSource, indent = 4)
+  implicit val rw: ReadWriter[SqlDataSourceConfig] = readwriter[Value].bimap[SqlDataSourceConfig](
```
[the remark this answers is not in the document]
Comment (pstutz) [reply spark-cypher/src/main/scala/org/opencypher/spark/api/io/sql/SqlDataSourceConfig.scala]: Yes, `Value` is a JSON value and uJson has built-in support for serialising sealed traits and adds a `$type` field to disambiguate different implementing classes. We merely add mappings that rename the field to `type` (and set nice values for that field with the `@key` annotation on the implementing classes).

**Pull request merged**